logstash-core/src/main/java/org/logstash/plugins/inputs/Stdin.java

```diff
@@ -0,0 +1,122 @@
+import co.elastic.logstash.api.Configuration;
+import co.elastic.logstash.api.Input;
+import co.elastic.logstash.api.LogstashPlugin;
+import co.elastic.logstash.api.Context;
+import co.elastic.logstash.api.PluginHelper;
+import co.elastic.logstash.api.PluginConfigSpec;
+import org.logstash.plugins.discovery.PluginRegistry;
+import org.logstash.execution.queue.QueueWriter;
+
+import java.io.FileDescriptor;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.nio.ByteBuffer;
+import java.nio.channels.AsynchronousCloseException;
+import java.nio.channels.FileChannel;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.function.Consumer;
+
+@LogstashPlugin(name = "java-stdin")
+public class Stdin implements Input, Consumer<Map<String, Object>> {
+
+    public static final PluginConfigSpec<String> CODEC_CONFIG =
+            Configuration.stringSetting("codec", "line");
+
+    private static final int BUFFER_SIZE = 64 * 1024;
+
+    private final LongAdder eventCounter = new LongAdder();
+    private String hostname;
+    private Codec codec;
+    private volatile boolean stopRequested = false;
+    private final CountDownLatch isStopped = new CountDownLatch(1);
+    private FileChannel input;
+    private QueueWriter writer;
+
+    /**
+     * Required Constructor Signature only taking a {@link Configuration}.
+     *
+     * @param configuration Logstash Configuration
+     * @param context       Logstash Context
+     */
+    public Stdin(final Configuration configuration, final Context context) {
+        this(configuration, context, new FileInputStream(FileDescriptor.in).getChannel());
+    }
+
+    Stdin(final Configuration configuration, final Context context, FileChannel inputChannel) {
+        try {
+            hostname = InetAddress.getLocalHost().getHostName();
+        } catch (UnknownHostException e) {
+            hostname = "[unknownHost]";
+        }
+        String codecName = configuration.get(CODEC_CONFIG);
+        codec = PluginRegistry.getCodec(codecName, configuration, context);
+        if (codec == null) {
+            throw new IllegalStateException(String.format("Unable to obtain codec '%a'", codecName));
+        }
+        input = inputChannel;
+    }
+
+    @Override
+    public void start(QueueWriter writer) {
+        this.writer = writer;
+        final ByteBuffer buffer = ByteBuffer.allocateDirect(BUFFER_SIZE);
+        try {
+            while (!stopRequested && (input.read(buffer) > -1)) {
+                buffer.flip();
+                codec.decode(buffer, this);
+                buffer.compact();
+            }
+        } catch (AsynchronousCloseException e2) {
+            // do nothing -- this happens when stop is called during a pending read
```

Comment:
That exception occurs in response to the user action of either shutting down LS or reloading the pipeline when the input loop is waiting on the `input.read(buffer)` statement rather than in the `codec.decode` statements (in practice, it's usually the former). I'm not sure that's useful information to log, but I'm open to reasons why it might be.

Reply:
If we have accepted bytes from the wire (e.g., `TCP ACK`), and those bytes don't end up contributing to events due to a shutdown or pipeline reload, having a clue somewhere in the logs could save a whole lot of debugging.